packages/stable/src/api/geospatial/types.ts

```diff
@@ -0,0 +1,245 @@
+  | 'POLYGONZ'
+  | 'MULTIPOINTZ'
+  | 'MULTILINESTRINGZ'
+  | 'MULTIPOLYGONZ'
+  | 'GEOMETRYM'
+  | 'POINTM'
+  | 'LINESTRINGM'
+  | 'POLYGONM'
+  | 'MULTIPOINTM'
+  | 'MULTILINESTRINGM'
+  | 'MULTIPOLYGONM'
+  | 'GEOMETRYCOLLECTIONM'
+  | 'GEOMETRYZM'
+  | 'POINTZM'
+  | 'LINESTRINGZM'
+  | 'POLYGONZM'
+  | 'MULTIPOINTZM'
+  | 'MULTILINESTRINGZM'
+  | 'MULTIPOLYGONZM'
+  | 'GEOMETRYCOLLECTIONZM';
+
+export type GeospatialPropertyType =
+  | 'STRING'
+  | 'LONG'
+  | 'DOUBLE'
+  | 'BOOLEAN'
+  | 'TIMESTAMP'
+  | GeometryPropertyType;
+
+export type GeospatialFeatureTypeProperty =
+  | {
+      type: Exclude<GeospatialPropertyType, GeometryPropertyType | 'STRING'>;
+      description?: string;
+      optional?: boolean;
+    }
+  | {
+      type: GeometryPropertyType;
+      srid?: number;
+      description?: string;
+      optional?: boolean;
+    }
+  | {
+      type: 'STRING';
+      size: number;
+      description?: string;
+      optional?: boolean;
+    };
+
+type Properties = Record<string, GeospatialFeatureTypeProperty>;
+
+export interface Geospatial {
+  featureTypeAPI: unknown;
+  featureAPI: unknown;
+}
+
+export interface GeospatialFeature extends ExternalId {
+  [property: string]:
+    | string
+    | number
+    | boolean
+    | Date
+    | { wkt: string }
+    | Geometry;
+}
+
+export interface GeospatialFeatureResponse extends GeospatialFeature {
+  createdTime: Date;
+  lastUpdatedTime: Date;
+}
+
+export interface FeatureType extends ExternalId, InternalId {
+  properties: Properties & {
+    _created_at: { type: 'LONG' };
+    _updated_at: { type: 'LONG' };
+    _external_id: { type: 'STRING'; size: 32 };
+  };
+  searchSpec?: {
+    [indexName: string]: {
+      properties:
+        | string[]
+        | ['_created_at']
+        | ['_updated_at']
+        | ['_external_id'];
+    };
+  };
+}
+
+type GeospatialIndexSpec = {
+  properties: string[];
+};
+
+export interface GeospatialCreateFeatureType extends ExternalId {
+  properties: Properties;
+  searchSpec?: {
+    [indexName: string]: GeospatialIndexSpec;
+  };
+}
+
+export interface GeospatialUpdateFeatureType extends ExternalId {
+  update:
+    | {
+        properties: { add: Properties } | { remove: string[] };
+      }
+    | {
+        searchSpec:
+          | {
+              add: { [indexName: string]: { properties: string[] } };
+            }
+          | { remove: string[] };
+      };
+}
+
+export interface GeospatialRecursiveDelete {
+  recursive?: boolean;
+}
+
+export interface GeospatialOutput {
+  output?:
+    | {
+        geometryFormat: GeometryType;
+      }
+    | {
+        properties: Record<string, unknown>;
+      };
+}
+
+interface GeospatialPropertyAndValue {
+  property: string;
+  value: unknown;
+}
+
+interface GeospatialPropertyAndPattern {
+  property: string;
+  pattern: string;
+}
+
+interface GeospatialAllowCrsTransformation {
+  allowCrsTransformation?: boolean;
+}
+
+interface RecursiveFilters {
+  or?: NonRecursiveFilters[];
```

Comment:
RecursiveFilters can have other Recursive filters as children, e.g.
```
{
  "and": [
      { "equals": ... },
      { "or": [..., ..., ] }
  ]
}
```
What is the reason that you distinguish between recursive and non-recursive filters?

Reply:
The distinguish is just for better breakdown.
Both are Recursive and Non-Recursive are local variables only and exported is complete filter only.

Reply:
but this breakdown is limiting the power of API (nested recursive filters)

Reply:
also, is plural for class/interface names a convention in JS?

Reply:
@pawelzwolenskicognite I've removed the recursive & non recursive.
Can you check if it seems correct now ?

Reply:
yes, much better, thanks!